qiskit/algorithms/optimizers/adam_amsgrad.py

```diff
@@ -260,6 +264,9 @@ def minimize(
             if self._snapshot_dir:
                 self.save_params(self._snapshot_dir)
 
+            if self.callback is not None:
+                self.callback(params_new)
```

Comment:
In scipy, the callback has a bool return (which I see the OptimizerCallback in Optimizer is typed that way too) if this returns True then the optimizer execution is terminated (you can see this in the callback parameter docs [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html) - perhaps we should be consistent with this here, and in other similar places, where it has been added.

Reply:
I am not sure I am following. You mean only trust-constr processes the bool and actually stops? This is a callback the user defines right, they get to see the values and if they return True from this routine then the execution stops. This is what the scipy callback docs say. So here I would expect the return of the `self.callback(params_new)` to be checked and if True for the optimizer execution to stop to be consistent with the expected behavior of the scipy ones.

Reply:
So rigorously, 

```python

@overload
def __call__(xk):
    ...
@overload
def __call__(xk, state: OptimizeResult) -> bool:
    ...

```
is correct signature. But I omitted the latter.